source/extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter.cc

```diff
@@ -0,0 +1,64 @@
+#include "extensions/filters/http/adaptive_concurrency/adaptive_concurrency_filter.h"
+
+#include <chrono>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "extensions/filters/http/adaptive_concurrency/concurrency_controller/concurrency_controller.h"
+#include "extensions/filters/http/well_known_names.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace AdaptiveConcurrency {
+
+AdaptiveConcurrencyFilterConfig::AdaptiveConcurrencyFilterConfig(
+    const envoy::config::filter::http::adaptive_concurrency::v2alpha::AdaptiveConcurrency&,
+    Runtime::Loader& runtime, std::string stats_prefix, Stats::Scope& scope,
+    TimeSource& time_source)
+    : runtime_(runtime), stats_prefix_(std::move(stats_prefix)), scope_(scope),
+      time_source_(time_source) {}
+
+AdaptiveConcurrencyFilter::AdaptiveConcurrencyFilter(
+    AdaptiveConcurrencyFilterConfigSharedPtr config, ConcurrencyControllerSharedPtr controller)
+    : config_(std::move(config)), controller_(std::move(controller)) {}
+
+AdaptiveConcurrencyFilter::~AdaptiveConcurrencyFilter() = default;
+
+Http::FilterHeadersStatus AdaptiveConcurrencyFilter::decodeHeaders(Http::HeaderMap&,
+                                                                   bool end_stream) {
+  if (!end_stream) {
```

Comment:
I think you need to check for forwarding whether it's end stream or not? Meaning, we should always check for forwarding when the request starts.